examples_tests/61.BlitFilterTest/main.cpp

```diff
@@ -0,0 +1,1561 @@
+		}
+
+		return result;
+	}
+
+private:
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> alphaTestDSLayout = nullptr;
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> alphaTestPipelineLayout = nullptr;
+
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> blitDSLayout = nullptr;
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> blitPipelineLayout = nullptr;
+
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> normalizationDSLayout = nullptr;
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> normalizationPipelineLayout = nullptr;
+
+	const uint32_t sharedMemorySize;
+	video::ILogicalDevice* device;
+
+	core::smart_refctd_ptr<video::IGPUSampler> sampler = nullptr;
+
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> getDSLayout(const uint32_t descriptorCount, const asset::E_DESCRIPTOR_TYPE* descriptorTypes, video::ILogicalDevice* logicalDevice, core::smart_refctd_ptr<video::IGPUSampler> sampler) const
+	{
+		constexpr uint32_t MAX_DESCRIPTOR_COUNT = 100u;
+		assert(descriptorCount < MAX_DESCRIPTOR_COUNT);
+
+		video::IGPUDescriptorSetLayout::SBinding bindings[MAX_DESCRIPTOR_COUNT] = {};
+
+		for (uint32_t i = 0u; i < descriptorCount; ++i)
+		{
+			bindings[i].binding = i;
+			bindings[i].count = 1u;
+			bindings[i].stageFlags = asset::IShader::ESS_COMPUTE;
+			bindings[i].type = descriptorTypes[i];
+
+			if (bindings[i].type == asset::EDT_COMBINED_IMAGE_SAMPLER)
+				bindings[i].samplers = &sampler;
+		}
+
+		auto dsLayout = logicalDevice->createGPUDescriptorSetLayout(bindings, bindings + descriptorCount);
+		return dsLayout;
+	}
+
+	static inline asset::E_FORMAT getCompatClassFormat(const asset::E_FORMAT format)
+	{
+		const asset::E_FORMAT_CLASS formatClass = asset::getFormatClass(format);
+		switch (formatClass)
+		{
+		case asset::EFC_8_BIT:
+			return asset::EF_R8_UINT;
+		case asset::EFC_16_BIT:
+			return asset::EF_R16_UINT;
+		case asset::EFC_32_BIT:
+			return asset::EF_R32_UINT;
+		case asset::EFC_64_BIT:
+			return asset::EF_R32G32_UINT;
+		case asset::EFC_128_BIT:
+			return asset::EF_R32G32B32A32_UINT;
+		default:
+			return asset::EF_UNKNOWN;
+		}
+	}
+
+	core::smart_refctd_ptr<asset::ICPUShader> loadShaderFromFile(const char* filepath)
+	{
+		auto sys = device->getPhysicalDevice()->getSystem();
+		system::ISystem::future_t<core::smart_refctd_ptr<system::IFile>> future;
+		sys->createFile(future, filepath, static_cast<system::IFile::E_CREATE_FLAGS>(system::IFile::ECF_READ | system::IFile::ECF_MAPPABLE));
+
+		const auto glslFile = future.get();
+		if (!glslFile)
+			return nullptr;
+
+		auto buffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(glslFile->getSize());
+		memcpy(buffer->getPointer(), reinterpret_cast<const system::IFile*>(glslFile.get())->getMappedPointer(), glslFile->getSize());
+
+		auto cpuShader = core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(buffer), asset::IShader::buffer_contains_glsl_t{}, asset::IShader::ESS_COMPUTE, "????");
+
+		return cpuShader;
+	}
```

Comment:
take a look at other utilities, they do it much faster and nicer (but your shaders need to be builtin resources)

Reply:
use the asset loader, `getAsset` dont do opening files by yourself.

Reply:
actually you don't need this function at all.

Reply:
Followed in the footsteps of `CScanner`.

Resolved?

Reply:
there is a better way https://github.com/Devsh-Graphics-Programming/Nabla/pull/343/files/90a68ce14ea09ca939b05feeb27adf38073708c8..4442fb70c98dead6b3be3eff2f6146bd0a466da5#r873930537

Reply:
Resolved.